### ISSUE TITLE
[Suggestion] Prediction on rank

### ISSUE THREAD

**Issue by NicolaosGreene** [KC3Kai/KC3Kai#728]
I would find it useful for some of the times when I am Orel cruising to tell if I am getting a B rank or not on a node during a day battle before deciding to proceed to night battle. However, determining if you are getting a B or not can be difficult in low damage situations if you don't know exact HPs of both you and the enemy since you need 2.5 times the percent of the damage they dealt.

**Comment by ReiFan49** [comment]
flagship sinking worth more,
losing a ship reduces one rank,
losing more than half of the fleet either D or C,

SS S A B vs C D E F
both of them should be calculated (flagship damage weight is greater)

**Comment by JD-Kurotoshigami** [comment]
this is pretty much the battle prediction itself, you just check #of sinked ships and the damage dealt and taken
though it would really help if it was already shown though

**Comment by nekoworkshop** [comment]
Here is a code snippet for rank calculation. Copied from https://osdn.jp/projects/kancollesniffer/, originally forked from https://github.com/sanaehirotaka/logbook

Havn't used kancollesniffer for a while but I do remember it managed to predict the rank just by api_req_sortie/battle.


``` 
       private BattleResultRank CalcResultRank()
        {
            var combined = _friend.Concat(_guard).ToArray();
            // Remaining ship after battle
            var friendNowShips = combined.Count(r => r.NowHp > 0);
            var enemyNowShips = _enemyHp.Count(hp => hp > 0);
            // Total damage
            var friendGauge = combined.Sum(r => r.Damage);
            var enemyGauge = _enemyStartHp.Sum() - _enemyHp.Sum();
            // K.O number
            var friendSunk = combined.Count(r => r.NowHp == 0);
            var enemySunk = _enemyHp.Count(hp => hp == 0);

            var friendGaugeRate = Floor((double)friendGauge / combined.Sum(r => r.StartHp) * 100);
            var enemyGaugeRate = Floor((double)enemyGauge / _enemyStartHp.Sum() * 100);
            var equalOrMore = enemyGaugeRate > (0.9 * friendGaugeRate);
            var superior = enemyGaugeRate > 0 && enemyGaugeRate > (2.5 * friendGaugeRate);

            if (friendSunk == 0)
            {
                if (enemyNowShips == 0)
                {
                    if (friendGauge == 0)
                        return BattleResultRank.P;
                    return BattleResultRank.S;
                }
                if (_enemyHp.Length == 6)
                {
                    if (enemySunk >= 4)
                        return BattleResultRank.A;
                }
                else if (enemySunk * 2 >= _enemyHp.Length)
                {
                    return BattleResultRank.A;
                }
                if (_enemyHp[0] == 0)
                    return BattleResultRank.B;
                if (superior)
                    return BattleResultRank.B;
            }
            else
            {
                if (enemyNowShips == 0)
                    return BattleResultRank.B;
                if (_enemyHp[0] == 0 && friendSunk < enemySunk)
                    return BattleResultRank.B;
                if (superior)
                    return BattleResultRank.B;
                if (_enemyHp[0] == 0)
                    return BattleResultRank.C;
            }
            if (enemyGauge > 0 && equalOrMore)
                return BattleResultRank.C;
            if (friendSunk > 0 && friendNowShips == 1)
                return BattleResultRank.E;
            return BattleResultRank.D;
        }
```

**Comment by ethrundr** [comment]
@ReiFan do not forget, if you loss all your ships (flagship taiha'd somehow is counted as "lost" in sortie), you'll got that dreaded blue E rank....

**Comment by nicky008** [comment]
Seem E rank is the forgotten rare rank in the game for now, as no one even try to do suicide in 100% completely lose.

**Comment by ReiFan49** [comment]
my friend have a video consisting the "how to get an E rank using 2 ship sortie"

**Comment by nicky008** [comment]
I had watch it a year ago, and some comments on that are so funny to watch.

**Comment by ethrundr** [comment]
@nicky008 nevertheless, we still need to take account of that rank if this feature is going to be implemented. rank is still a rank when exist in the system after all.

**Comment by ReiFan49** [comment]
![image](https://cloud.githubusercontent.com/assets/9847780/9829780/ddf15e72-593d-11e5-8b81-ee8d8af4418f.png)
first time kuso. EVER.

**Comment by ethrundr** [comment]
wohooo \ :laughing: /
that dreaded blue E rank

**Comment by Javran** [comment]
I'm interested in this. what's the status? with battle prediction enabled, we can display a predicted rank, maybe adding some indicator to make it looks different than the actual rank.

**Comment by dragonjet** [comment]
hm we could just use the same icons, but with a small (?) on the side, and slightly transparent. When the actual results come in, it will go back to normal without ? and full opaque.

Though I don't think anyone is working on it

**Comment by Javran** [comment]
@dragonjet that sounds good to me, I may come back working on this after the goal-template stuff is done.

**Comment by Yukinyaa** [comment]
I think this is hi priority feature for many.

**Comment by dragonjet** [comment]
I see it only useful for 1-1 though

**Comment by Yukinyaa** [comment]
Many thinks it is very useful for rank point cruising

**Comment by dragonjet** [comment]
>some that says they will use kc3kai when they have this feature

alright then. this next priority jk :laughing:

**Comment by Javran** [comment]
@dragonjet are you working on this?

**Comment by dragonjet** [comment]
oops, I just saw this after nearly 2 mos, nope I'm not working on it atm, but I see you made `dev-rank-predict`. Whoever wants to do continue this may continue it on that branch. thanks :+1:

**Comment by sinsinpub** [comment]
Ive tried this branch... is there just missed the presentation of predicated rank?
I have added the predicated rank image icon to the same position `.module.activity .battle_rating` (natsuiro theme now), just with `opacity: 0.5`, for both day and night.
And on battle result, real rank will be shown with 100% opacity.

btw: the KO stamp really not work for now. just a red line left :joy:

**Comment by ReiFan49** [comment]
uhh... the KO stamp and that E-rank for :rocket: reminder .... (stares the sinking post)

**Comment by sinsinpub** [comment]
An issue report for current impl:
For night battle only node (eg: 2-5-D), it seems no prediction result built.

**Comment by dragonjet** [comment]
Merged #1263
